### ISSUE TITLE
flaky TestSingleRelay

### ISSUE THREAD

**Issue by marten-seemann** [libp2p/go-libp2p#1403]
```
  === RUN   TestSingleRelay
  2022-04-16T15:29:07.869Z	DEBUG	autorelay	autorelay/relay_finder.go:540	starting relay finder
  2022-04-16T15:29:07.971Z	DEBUG	autorelay	autorelay/relay_finder.go:169	found node	{"id": "QmTbfTMXwfj6U58BNAPxCaDMfgGoonCHv5YNjg4tYwDfkD"}
  2022-04-16T15:29:08.015Z	DEBUG	autorelay	autorelay/relay_finder.go:169	found node	{"id": "QmfW7CeX8scH6vLyTFofQxzznXvjVxpaHp8DSEbERvEYa8"}
  2022-04-16T15:29:08.047Z	DEBUG	autorelay	autorelay/relay_finder.go:219	node supports relay protocol	{"peer": "QmfW7CeX8scH6vLyTFofQxzznXvjVxpaHp8DSEbERvEYa8", "supports circuit v2": true}
  2022-04-16T15:29:08.052Z	DEBUG	autorelay	autorelay/relay_finder.go:371	adding new relay	{"id": "QmfW7CeX8scH6vLyTFofQxzznXvjVxpaHp8DSEbERvEYa8"}
  2022-04-16T15:29:08.080Z	DEBUG	autorelay	autorelay/relay_finder.go:219	node supports relay protocol	{"peer": "QmTbfTMXwfj6U58BNAPxCaDMfgGoonCHv5YNjg4tYwDfkD", "supports circuit v2": true}
  2022-04-16T15:29:08.083Z	DEBUG	autorelay	autorelay/relay_finder.go:371	adding new relay	{"id": "QmfW7CeX8scH6vLyTFofQxzznXvjVxpaHp8DSEbERvEYa8"}
  2022-04-16T15:29:08.090Z	DEBUG	autorelay	autorelay/relay_finder.go:371	adding new relay	{"id": "QmTbfTMXwfj6U58BNAPxCaDMfgGoonCHv5YNjg4tYwDfkD"}
  2022-04-16T15:29:08.133Z	DEBUG	autorelay	autorelay/relay_finder.go:169	found node	{"id": "QmcHuH5KAC1aMdhCGSrwZ8w7rtouVh58HN1Qy4pLaSLETo"}
  2022-04-16T15:29:08.133Z	DEBUG	autorelay	autorelay/relay_finder.go:174	skipping node. Already have enough candidates	{"id": "QmcHuH5KAC1aMdhCGSrwZ8w7rtouVh58HN1Qy4pLaSLETo", "num": 2, "max": 1}
  2022-04-16T15:29:08.431Z	DEBUG	autorelay	autorelay/relay_finder.go:169	found node	{"id": "QmaZp9dd2ckbWGARNqDqLSCUvuDNLh3JYskEqjMCSkcZth"}
  2022-04-16T15:29:08.431Z	DEBUG	autorelay	autorelay/relay_finder.go:174	skipping node. Already have enough candidates	{"id": "QmaZp9dd2ckbWGARNqDqLSCUvuDNLh3JYskEqjMCSkcZth", "num": 2, "max": 1}
  2022-04-16T15:29:08.599Z	DEBUG	autorelay	autorelay/relay_finder.go:169	found node	{"id": "QmehaW9LCzd5oQN6A5RMQNcAbnYB628h52PpUKFeYxoJXu"}
  2022-04-16T15:29:08.599Z	DEBUG	autorelay	autorelay/relay_finder.go:174	skipping node. Already have enough candidates	{"id": "QmehaW9LCzd5oQN6A5RMQNcAbnYB628h52PpUKFeYxoJXu", "num": 2, "max": 1}
      autorelay_test.go:147: 
          	Error Trace:	autorelay_test.go:147
          	Error:      	Condition satisfied
          	Test:       	TestSingleRelay
  2022-04-16T15:29:08.645Z	DEBUG	autorelay	autorelay/relay_finder.go:552	stopping relay finder
  --- FAIL: TestSingleRelay (1.22s)

**Comment by marten-seemann** [comment]
The problem here seems to be that `handleNewCandidate` can be called multiple times and concurrently.

**Comment by marten-seemann** [comment]
Fixed by #1404.
